### ISSUE TITLE
[RFC] Switch to JSON

### ISSUE THREAD

**Issue by digama0** [leanprover/lean4export#3]
The concrete syntax of the export format is troublesome to parse, being a homegrown textual format with a heavy reliance on newline separators and with lots of "clever" sequence encodings. Moreover, the encoding of names is unquoted, which is just plain wrong in the presence of names with escaped characters, because these names can include arbitrary characters, including newlines, keywords and everything else - a classic SQL injection attack.

I propose we drop this ad hoc encoding entirely and switch to a JSON-based format. This is much easier to get right, and libraries for doing the parsing are numerous (but it's also feasible to write the parser directly).

**Comment by ammkrn** [comment]
Fwiw I'm not opposed to the idea; while the current format is really easy to write a parser for, JSON would certainly be more accessible for people/projects that don't want to write a parser in the first place. I haven't looked deeply into the name escaping issue yet, but I'll take your word for it that it's a potential source of problems. 

I'm interested to see what you come up with.

**Comment by adomasbaliuka** [comment]
JSON has a quite complicated spec with lots of extensions which aren't implemented exactly the same (or strictly according to the spec) by different libraries. 

Perhaps something slightly more limited and more standardized would be better? (Ideally one would also use a formally verified parser, which after searching for a bit I'm still not quite sure about the status of for JSON)

Perhaps a subset of JSON would work? (Which one would specify separately for the purpose of lean4export.)

**Comment by digama0** [comment]
I'm not sure what you are referring to, JSON is an exceedingly simple format. The only part I'm aware of tricky bits in json is for representation of large integers and floating point, but I don't think these are likely to be issues since we don't need floats at all and it is unlikely for numbers to get that large except when representing lean bignums, and string encoding these solves the problem completely.

**Comment by adomasbaliuka** [comment]
All I'm saying is that there isn't an ironclad standard for how to parse JSON and many (also "standard") parser implementations differ. Some examples: [here](https://seriot.ch/projects/parsing_json.html).

I thought maybe it's important for this export format to have an ironclad standard. Am I wrong? I guess it's also reasonable to say something like "let the parser fail on exported valid lean proofs in extremely rare cases (e.g. if they contain too big numbers), let it crash on malformed input, let it output whatever it wants, the kernel will check everything in the end and that's all that matters".

**Comment by digama0** [comment]
> All I'm saying is that there isn't an ironclad standard for how to parse JSON and many (also "standard") parser implementations differ. Some examples: [here](https://seriot.ch/projects/parsing_json.html).

This is not true. What that page shows is that there exist multiple documents that specify JSON, and parsers sometimes accept more or less than the spec due to implementation limits or "extensions". I don't see why any of this matters, I think you should be more specific.

Regarding having an ironclad standard, I'm not seeing anything here which prevents having such. But this is a proof format, which means that it actually doesn't matter if there are edge cases which are interpreted oddly, because proof checkers are allowed to spuriously fail for implementation limits reasons or even just not liking the shape of the proof. That's a quality-of-implementation issue, not a correctness issue.

**Comment by daniel-levin** [comment]
> I thought maybe it's important for this export format to have an ironclad standard. Am I wrong?

No, you are not wrong. I don't think anyone wants to allow spurious errors to be introduced by a poorly designed serialization format.

With this shared goal in mind, I maintain that imperfect serialization formats, such as JSON, are fine for all pragmatic purposes, provided their shortcomings are dealt with by the serializer - in this case lean4export.  

Moreover, "ironclad" serialization formats are rarely truly ironclad. Even the most strictly specified, such as BER/DER - widely used in public key cryptography systems - are replete with deliberately non-conformant  implementations. Example: https://github.com/RustCrypto/formats/issues/779.

Finally, the advantages of an export format that can be understood by practically every programming language far outweigh the downsides. Sure, using JSON makes ambiguity and/or invalidity possible, though unlikely. The principal advantage of JSON (or TOML, my preferred format) is ease of implementation. I am optimistic that making it easy to get started will result in more implementations, which will result in more scrutiny of the proof terms constructed by Lean. A more diverse set of kernels will necessarily result in more scrutiny of the serializer itself. By this logic, JSON is a reasonable candidate.

**Comment by ammkrn** [comment]
The issue of escaped identifiers is now a live issue, I think due to these `library_note` things, e.g. [this](https://github.com/leanprover-community/mathlib4/blame/f4dee013552444b4b058a624930590186361ac8e/Mathlib/Algebra/HierarchyDesign.lean#L204) produces a `NS` export line with a space in the name suffix.

**Comment by ammkrn** [comment]
I'm going to be moving forward with this since it's currently an issue in mathlib as mentioned previously; here's an initial pass for what the object layout might look like [as markdown](https://gist.github.com/ammkrn/1649110f70005ab54e280bbe26110b87) and as an (llm derived)  [json spec](https://gist.github.com/ammkrn/d68a78873c9e48e6ec4e971efc230a56). I'm assuming the norm will be ndjson.

This follows Lean's rules for JSON serialization of inductives (the ones used by the deriving handlers) with the exception that one layer of `val` bureaucracy for the `ConstantVal` items is removed.

**Comment by digama0** [comment]
LGTM

**Comment by Gravifer** [comment]
What about migrating to something like [EDN](https://en.wikipedia.org/wiki/Clojure#Extensible_Data_Notation)? I don't think you can bypass parser writing that way (much smaller userbase), but it does have better type semantics; should be less ambiguous

**Comment by ammkrn** [comment]
> What about migrating to something like [EDN](https://en.wikipedia.org/wiki/Clojure#Extensible_Data_Notation)? I don't think you can bypass parser writing that way (much smaller userbase), but it does have better type semantics; should be less ambiguous

My assumption is that people will want to stick with JSON unless you're aware of any existing ambiguity concerns that might be an issue (if so, please share them). On the bright side the exporter is very small, so it should be easy to fork in order to support different formats.

**Comment by nomeata** [comment]
How concrete is this proposal here? Do we already know how much larger exports of, say, mathlib get, and whether parsing is slower or faster?

**Comment by ammkrn** [comment]
> How concrete is this proposal here? Do we already know how much larger exports of, say, mathlib get, and whether parsing is slower or faster?

At full verbosity using [this](https://github.com/ammkrn/lean4export/tree/json_output) the output is basically twice the size. I'm not opposed to having a "compact" flag for using shortened property/field names. I don't speak for Lean or the FRO, but since this is the semi-official export tool, I would assume the desire is for something legible that's not going to break, at least as the default output.

> How concrete is this proposal here?

I think we're all open to suggestions if you have some.

**Comment by nomeata** [comment]
My gut feeling likes the old format, but for no good reason, and factor 2 seems reasonable, so no complaints from my side.

The main reason I'm asking is if I'd build tools that consume the exported format, if I should simply ignore the old format because it's going to be replaced real soon now, or if this is more a mid-term project here and I shouldn't wait for it.

**Comment by ammkrn** [comment]
> My gut feeling likes the old format, but for no good reason, and factor 2 seems reasonable, so no complaints from my side.
> 
> The main reason I'm asking is if I'd build tools that consume the exported format, if I should simply ignore the old format because it's going to be replaced real soon now, or if this is more a mid-term project here and I shouldn't wait for it.

You should probably ignore the old (current) format, because it's incapable of exporting mathlib releases after the introduction of library_note2, and there are no plans to patch it. In addition to the identifier/string issue, it's not really capable of exporting mdata expressions which at least one user has credibly expressed interest in doing.

The fork I linked in my previous message should more or less work right now. I need to finish writing a new parser, probably make some changes to the exporter in response to whatever is going to come up in testing, and then I'll file a PR.

**Comment by nomeata** [comment]
What is your plan for large integer literals? The spec draft above has `integer` as the JSON type; this requires JSON parsers to be able to handle bignum literals. Is this easily possible in typical off-the-shelve JSON parsers?

A [common alternative is to serialize them as JSON strings](https://www.hackerone.com/blog/safely-handling-large-integers-json-best-practices-and-pitfalls), it seems. (No preference on my side, as long as its considered and documented)

**Comment by ammkrn** [comment]
> What is your plan for large integer literals? The spec draft above has `integer` as the JSON type; this requires JSON parsers to be able to handle bignum literals. Is this easily possible in typical off-the-shelve JSON parsers?
> 
> A [common alternative is to serialize them as JSON strings](https://www.hackerone.com/blog/safely-handling-large-integers-json-best-practices-and-pitfalls), it seems. (No preference on my side, as long as its considered and documented)

Thanks, I did run into this while updating downstream. There's now an option `--nat-format=<number | le-bytes | string>` that controls how nat literal values are serialized.

[This branch](https://github.com/ammkrn/lean4export/tree/json_output) is up to speed if you want to take a look, the option is documented in the README. The README needs a little more work before the PR, if nothing else I need to revisit the initial spec and make some updates, then fix the link.

One thing you might know @nomeata is if there's a reasonable way to reference the current package version as specified in the lakefile from within the project itself; the exporter outputs a "metadata" object that has info about the exporter version, and it would be nice to be able to reuse the value in the lakefile rather than having the value exist in potentially two places (the lakefile and in the project's executable code). Or if you just have a better solution generally I'm all ears.

**Comment by nomeata** [comment]
Is it a good idea to have an option? We want the export format to be simple and universal, and be able to process an exported proof with all available external checkers. It's be silly to have to export it multiple times to check with different checkers.

I'd suggest to just make a judgment call and stick to it. With my limited knowledge of the JSON library ecosystem I'd lean towards using strings, because it's less likely to cause issues.


--- 

Re embedding versions:
I don't know if that feature exists in lake, probably not. Maybe it could be implemented using dynamic code in the lakefile.lean, but not sure if I'd recommend it.

**Comment by ammkrn** [comment]
Fair enough. I've removed the option and specialized to `string`.

**Comment by nomeata** [comment]
Now that we had a few people look and work with at the new format, maybe do a round of final polishing? I believe @hargoniX has some thoughts.

From my side I'd prefer a less verbose format. Not only to be not unnecessary wasteful with disk space, but also because I find it easier to read when I have to look at it. So I’d propose (without insisting on it)

* shorter field names: `name` and `type` rather than `binderName` and `binderType`.

* Immediate data for unary constructors:
  ```
  {"bvar":4,"i":1000}
  {"sort":5,"i":1000}
  {"lit":{"nat":5},"i":1000}  -- or even:  {"natLit":5,"i":1000}
  ```

**Comment by ammkrn** [comment]
I'm not opposed to abbreviated names (though I did try it out early on and was surprised by how little space it saved). The nat and string literals do currently have immediate data, since the "faithful" JSON representation would have had them both begin with the same `"lit"` json property which would have been troublesome for parsing.

There seems to be agreement that `declName`, `binderName`, and `deBruijnIndex` are on the chopping block (with bvar just being flattened to `{ "bvar": integer, "i": integer}` altogether. If there are any more identifiers on the list, let's hear 'em.

**Comment by nomeata** [comment]
Great!

Another suggestion:

You currently use `"i":123` for all references, but in fact these have different namespaces, and when manually inspecting the file it’s a bit tedious that I can not just search for `"i":263677` to find that name, as it will also find the expression with that identifier.

This might be a bit easier if the objects used `"ie"`, `"in"` and `"il"` to indicate in which namespace they live.

**Comment by hargoniX** [comment]
Two of other things I noticed:
- why do Level.max/imax take their arguments as an array as opposed to everything else?
- I do not understand how to recover an mdata node with the currently used serialization mechanism. The resulting JSON seems entirely unrecoverable to the original format to me in the general case.

**Comment by nomeata** [comment]
> why do Level.max/imax take their arguments as an array as opposed to everything else?


I noticed that as well, but didn't find it too strange: these have arguments that are clearly positional (left and right), while the others have arguments with no particular semantically meaningful order.

**Comment by ammkrn** [comment]
> Two of other things I noticed:
> 
>     * why do Level.max/imax take their arguments as an array as opposed to everything else?
> 
>     * I do not understand how to recover an mdata node with the currently used serialization mechanism. The resulting JSON seems entirely unrecoverable to the original format to me in the general case.

Perhaps I should have disclosed earlier, but I have no idea what the proposed use cases are for `mdata` in the export file and didn't think very hard about what to do with it. I have no objection to changing the layout and have no input on what a better format might be, other than maybe using the same integer based referencing for the name part of the KVMap keys.

max/imax use an array because they're the only elements that have anonymous constructor arguments and I was (at least initially) trying to stick to what Lean was doing.

**Comment by digama0** [comment]
I think it would be best to keep `mdata` because it changes the expr hash. There was recently a change to remove `nonDep` from export to fix the same issue there but I don't think stripping mdata from oleans is reasonable, so I think the checker should just support mdata and do nothing with it (treat it like an identity function which eagerly unfolds in whnf).

**Comment by nomeata** [comment]
> I think the checker should just support mdata and do nothing with it (treat it like an identity function which eagerly unfolds in whnf).

right, which is also what the official kernel does.

OTOH, it seems a bit overkill to define a stable format for the whole algebra of mdatas, including KVMap and Syntax(!), when for the envisioned applications of the export format we have no use for this. This also means that the format will not be stable across version of lean that only change the MData. And, at least ideally, the expression hash is an implementation detail that shouldn’t have relevance for an external stable proof representation.

**Comment by ammkrn** [comment]
I can deal with mdata potentially being included in the default output, but if that's going to be the case I would really prefer to retain a flag to get an export with no mdata. FWIW I do not intend to support mdata in the rust checker.

**Comment by nomeata** [comment]
@ammkrn, from reading your rust parser it seems that you assume that the indices in the JSON file are used in a strictly increasing contiguous fashion. Is that an invariant you wanted to be part of the format, or was that just an implemenation shortcut in the parser for now? 
(Requiring them to be in increasing order may simplify some export-to-export-format transformation tools, e.g. tools that filter the environment.)

**Comment by ammkrn** [comment]
Off the top of my head I'm pretty sure I can make changes to accommodate indices that appear out of order, but non-contiguous might be tougher. The current setup arises pretty naturally out of the way the exporter currently works and is a nice sanity check for consumers, but maybe you had some kind of parallelism in mind?

I do have one more change I'd like to push that tries to ensure the elements of an inductive declaration (inductive, constructors, recursors) are exported together.

It may or may not also be worth it to have something that tries to export the relevant `Nat` and `String` declarations on the first appearance of a nat or string literal. It's technically possible to export an `Expr.lit _` before the appropriate types are declared; users may expect that in any lean module using a literal (in a project that also has the requisite Nat/String declarations *somewhere*) that they're always going to get the "fast" kernel behavior, but that's not necessarily the case.

**Comment by nomeata** [comment]
I'm hesitant to make such rules part of the export format requirements. It seems that checkers (at least lean4checker and lean4lean) already do a topological sort after parsing.

I think it's reasonable to have a literal without any further operations in an export, at least if you don't do any operations with it. 

This is probably bike shedding dangerous corner, given that it's mostly relevant for artificially short exports.

**Comment by ammkrn** [comment]
I agree that it's reasonable for an exported environment to have a literal without anything else, the idea is that if environment to be exported has the relevant String/Nat declarations AND they are able to be exported prior to any Expr.lit items, it might be advantageous for the exporter to do that.

> I'm hesitant to make such rules part of the export format requirements. It seems that checkers (at least lean4checker and lean4lean) already do a topological sort after parsing.

To the extent this is not about the literals, in the interest of helping any consumer software stay as small as possible, I think when we have to answer the question of "should the exporter be responsible for this or the type checker", the default answer should be the exporter. Even if the type checker has to do something to ensure an invariant/order/whatever has been done properly, it's generally going to be smaller than if it had to do the thing itself.

I can change the export format for inductives to reflect the fact that they should be exported together, I think something like this is probably better anyway, especially for complex mutual/nested declarations:
```
{
    "inductive": {
        "inductInfos": [..]
        "ctorInfos": [..]
        "recInfos": [..]
    }
}
```

**Comment by nomeata** [comment]
If we are going that direction then one could argue that really the entries of the export should be modeled after [Lean.Declaration](https://leanprover-community.github.io/mathlib4_docs/Lean/Declaration.html#Lean.Declaration), and really reflect what’s being passed to the kernel, and not what comes out of it. For example, there should be no `recInfo`, as that’s something that the kernel produces based on the `inductDecl`.

**Comment by nomeata** [comment]
But that’s turning into a much bigger change than just the replacement of the textual format

**Comment by nomeata** [comment]
JFTR, there was some discussion over at https://github.com/leanprover/lean4export/pull/10#issuecomment-3744831891 where @ammkrn made good arguments why the slighly redundant format is better. Instead of going from `List Lena.ConstantInfo` to `List Lean.Declaration` all the way, his proposed format is a combination of the two that combines the best of both worlds, which is convenient for different consumers.

**Comment by nomeata** [comment]
Looks like we are converging?

What about versioning? The proposed metadata has fields for tool versions, but should we add a field for the format version itself, so that future changes to the format can get a version, and parsers can implement backwards compatibility?

**Comment by nomeata** [comment]
@hargoniX, any final suggestions for this iteration of the format? Will you merge?

**Comment by hargoniX** [comment]
I think it is fine yes.
